src/main/common/vector.h

```diff
@@ -26,124 +26,44 @@
        float x, y, z;
     };
-} fpVector3_t;
+} vector3_t;
 
-typedef struct {
+typedef struct matrix33_s {
     float m[3][3];
-} fpMat33_t;
-
-static inline fpVector3_t * vectorZero(fpVector3_t *v)
-{
-    v->x = 0.0f;
-    v->y = 0.0f;
-    v->z = 0.0f;
-    return v;
-}
-
-static inline float vectorNormSquared(const fpVector3_t * v)
-{
-    return sq(v->x) + sq(v->y) + sq(v->z);
-}
-
-static inline float vectorNorm(const fpVector3_t * v)
-{
-    return sqrtf(vectorNormSquared(v));
-}
-
-static inline fpVector3_t * vectorCrossProduct(fpVector3_t *result, const fpVector3_t *a, const fpVector3_t *b)
-{
-    fpVector3_t ab;
-
-    ab.x = a->y * b->z - a->z * b->y;
-    ab.y = a->z * b->x - a->x * b->z;
-    ab.z = a->x * b->y - a->y * b->x;
-
-    *result = ab;
-    return result;
-}
-
-static inline fpVector3_t * vectorAdd(fpVector3_t *result, const fpVector3_t *a, const fpVector3_t *b)
-{
-    fpVector3_t ab;
-
-    ab.x = a->x + b->x;
-    ab.y = a->y + b->y;
-    ab.z = a->z + b->z;
-
-    *result = ab;
-    return result;
-}
-
-static inline fpVector3_t * vectorScale(fpVector3_t *result, const fpVector3_t *a, const float b)
-{
-    fpVector3_t ab;
-
-    ab.x = a->x * b;
-    ab.y = a->y * b;
-    ab.z = a->z * b;
-
-    *result = ab;
-    return result;
-}
-
-static inline fpVector3_t * vectorNormalize(fpVector3_t *result, const fpVector3_t *v)
-{
-    float normSq = vectorNormSquared(v);
-    if (normSq > 0) {              // Hopefully sqrt(nonzero) is quite large
-        return vectorScale(result, v, 1.0f / sqrtf(normSq));
-    } else {
-        return vectorZero(result);
-    }
-}
-
-static inline fpVector3_t * matrixVectorMul(fpVector3_t * result, const fpMat33_t * mat, const fpVector3_t * a)
-{
-    fpVector3_t r;
-
-    r.x = mat->m[0][0] * a->x + mat->m[0][1] * a->y + mat->m[0][2] * a->z;
-    r.y = mat->m[1][0] * a->x + mat->m[1][1] * a->y + mat->m[1][2] * a->z;
-    r.z = mat->m[2][0] * a->x + mat->m[2][1] * a->y + mat->m[2][2] * a->z;
-
-    *result = r;
-    return result;
-}
-
-static inline fpVector3_t * matrixTrnVectorMul(fpVector3_t * result, const fpMat33_t * mat, const fpVector3_t * a)
-{
-    fpVector3_t r;
-
-    r.x = mat->m[0][0] * a->x + mat->m[1][0] * a->y + mat->m[2][0] * a->z;
-    r.y = mat->m[0][1] * a->x + mat->m[1][1] * a->y + mat->m[2][1] * a->z;
-    r.z = mat->m[0][2] * a->x + mat->m[1][2] * a->y + mat->m[2][2] * a->z;
-
-    *result = r;
-    return result;
-}
-
-static inline float vector2Cross(const fpVector2_t *a, const fpVector2_t *b)
-{
-    return a->x * b->y - a->y * b->x;
-}
-
-static inline float vector2Dot(const fpVector2_t *a, const fpVector2_t *b)
-{
-    return a->x * b->x + a->y * b->y;
-}
-
-static inline float vector2Mag(const fpVector2_t *a)
-{
-    return sqrtf(sq(a->x) + sq(a->y));
-}
+} matrix33_t;
+
+void vector2Zero(vector2_t *v);
+void vector2Add(vector2_t *result, const vector2_t *a, const vector2_t *b);
+void vector2Scale(vector2_t *result, const vector2_t *a, const float k);
```

Comment:
What's the reason for changing the  return value to void for these functions (and the other functions returning a vector). I found it convenient to have them return a pointer when working on the IMU code.
Does it trigger some warnings/errors about unused return values?

Reply:
It lets you return a pointer to memory which might not be allocated anymore in special cases. For example

```c
vector3_t *doSomething()
{
    vector3_t v1;
    return vector2Zero(&v1);
}

void doSomethingGreater()
{
    // This is undefined behavior because the memory, which
    // doSomething() is pointing to is already freed and doesn't
    // exist anymore from the firmware's POV.
    float value = doSomething()->x;  
}
```

Maybe this example is a bit too simplistic (because who writes nonsenical stuff like this) but I hope you get the idea. Functions returning pointers to memory you want to operate on is permitted on purpose in BF so that you cannot run into this problem at all, even if you try. It improves the resilience of the firmware against undefined behavior. NASA, for example, as many others utilize this rule as well, to write flight-proof and space-proof code.

I know it's inconvenient for vector/matrix operations but at the same time you "don't even need to know what you are doing" to not run into this issue randomly. Especially because BF is open-source and therefore a lot of people with different skill levels work on the codebase. It eliminates a potential "tripping hazard" of this library.

Reply:
> Functions returning pointers to memory you want to operate on *is permitted* on purpose in BF so that you cannot run into this problem at all, even if you try. 
That should be *is not permitted*, right? If so, I get it.

Reply:
yes, sorry. I mixed it up with "prohibited / not permitted" 😃